### PR TITLE
Moves atmos checking from Fire Alarms to Fire Locks, Fire Locks now close on their own rather than area-based, a bunch of related stuff.

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
- Moves hot/cold checks to the fire locks
- Makes detection due to above only close the affected Fire Lock and its neighbors (via merge code stolen from tanks)
- Fire Alarms still can lock an area down if clicked by a user, right clicking is now the reset function. Resetting also resets all Fire Locks closed by their own detection.
- Gives Fire Locks an overlay of lights that display the reason it closed (Orange for Hot, Cyan for Cold, White for unknown/Fire Alarm Pulled)
- Emagging a Fire Alarm now emags all Fire Locks in the area, which accomplishes the same effect as before (disables atmos monitoring). Pulling a Fire Alarm will make emagged Fire Locks in the area display a corrupted warning light pattern, making finding and replacing emagged Fire Locks somewhat easier.
- Silicons can disable atmos sensors in an area by CTRL-Clicking a Fire Alarm, much like before. but now the Fire Alarms in the area will change their sprite to hint at this (the normally-green light will turn off).

## Why It's Good For The Game
Makes a breach or fire not cause Fire Lock Hell anymore. Only the Fire Locks facing an issue will tend to close, leaving the rest of the area mostly unaffected.

## Changelog

:cl:
balance: Fire Locks now check for atmos issues on their own, rather than leaving it to the Fire Alarm. This should cut down on a single breach turning the main halls into a navigational mess.
balance: Fire Locks will wait five seconds after a reset before triggering again.
tweak: Fire Alarms will still activate all Fire Locks in an area if manually pulled.
tweak: Fire Alarms are still used to reset Fire Locks, but now it's a right-click action.
tweak: Emagging a Fire Alarm will apply the effect to all Fire Locks in its area, behaving roughly the same as before.
tweak: Disabling Fire Detect is also still done at the Fire Alarm and should work closely as it used to.
balance: Right-clicking a closed Fire Lock with a crowbar will open it for exactly two seconds. Holding it open (left click) is unchanged.
balance: Firelock emag functions (ignoring atmos detect) is now done by a doorjack on the firelock itself. You can still emag a fire alarm to mute it, though.
sprite: Fire Locks now have lights that display the reason they are active. Orange is Hot, Cyan means Cold, and White means a Fire Alarm was pulled.
/:cl:

~~I'm not done yet, but I figured I'd give people a chance to tell me how shit my code is while I finish up.~~

- [x] Remove all DEBUG tags after fixing the code they're next to
- [x] Re-test everything

Have a video of what the lights look like.
https://user-images.githubusercontent.com/37497534/136741583-632da9d9-913d-4e55-b468-f3490ea2580e.mp4

### CODE REVIEW

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Hard to tell what code's gonna stick around, and what won't, so I won't complain about that. Mostly structural stuff, especially those signals

**Comment by Ghilker** [comment]
Good work, only thing I think should change is that the fire doors should check the open turfs nearby for atmos changes and not the turf they are on for open/close behaviour
This way they could reopen automatically when the threat passed on both sides

**Comment by Colovorat** [comment]
> There's two reasons I don't want to do auto-reopen (yet).
> 
> The primary reason is that I'd need to set up checks for the four adjacent tiles, and then update them when their state changes (or else the firelock will never disable if an adjacent firelock closes and traps a tile of cold gas). There's just a lot of overhead adding to an already long PR
> 
> The secondary reason is that if there's an oscillating gas change for a particular tile (say, a breach adjacent to a vent), you could end up with a firelock that constantly cycles. Since we check based on temp rather than atmos pressure, this _probably_ isn't a likely scenario, but I'd want to test it all first.

rather checking adjacent tiles check for fire alarms? the we get... the same system we have now... huh

**Comment by LemonInTheDark** [comment]
> I don't agree with this change, minor inconveniences are a necessary part of simulating player agency (by putting obstacles in their way).
> 
> If you go around removing or shortening any minor inconveniences because as an experienced player you've dealt with them many times and they're starting to bore you you're actually harming the feeling of player agency and disconnecting yourself further from the game.
> 
> This kind of QoL development is an actively harmful way of contributing to the project.

To be honest with you a change along these lines is something I've wanted for a while, mostly because of the issues that making firelocks react to cold has brought.
Especially in large areas a single breach can cause aggressive levels of alarm fatigue, so much so that people will (at least when I was keeping a close eye on it) deconstruct all fire alarms, or just knock holes through a hallway of firelocks. Ends up just leading to a station wide cold snap, but it doesn't really seem to register as a bad thing.

I've never really thought about it in the way you're describing. It's honestly a bit hard for me to gauge my own thoughts like that. People being upset is certainly not a bad thing by default, I guess my instinct is to try and correct the behavior I've seen by making the things more targeted, instead of being a department or "hallway loop" wide thing. Maybe that's being overly prescriptive though

**Comment by optimumtact** [comment]
Turn the atmos tech role into another cargo tech then, see if I care.

**Comment by zxaber** [comment]
> Turn the atmos tech role into another cargo tech then, see if I care.

I'm not removing the need to fix breaches or fix air in rooms after. But a single missing window need not slow down foot traffic across the entire station

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Initial review, missing some stuff. Sorry bout the tardyness/volume

File: code/game/machinery/doors/firedoor.dm [hunk 1]
```diff
@@ -44,29 +76,142 @@
...
+	for(var/area/place in affecting_areas)
+		LAZYADD(place.active_firelocks, src)
+		if(LAZYLEN(place.active_firelocks) == 1) //if we're the first to activate in this particular area
+			for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+				fire_panel.set_status()
+			if(place == my_area)
+				place.alarm_manager.send_alarm(ALARM_FIRE, place) //We'll limit our reporting to just the area we're on. If the issue affects bordering areas, they can report it themselves
+	update_icon() //Sets the lights even if the door doesn't move.
+	if(!being_held_open)
+		INVOKE_ASYNC(src, .proc/correct_state)
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:190]: Check has been moved in the next commit.

File: code/game/machinery/doors/firedoor.dm [hunk 2]
```diff
@@ -44,29 +76,142 @@
...
+
+/**
+ * Proc that handles reset steps
+ *
+ * Sets the alarm_type to null, removes us from active firelock lists
+ * in the affected areas, and may tell fire alarms to update their icon
+ * Also sets the detect cooldown so that we can't immediately re-activate
+ * and finally calls correct_state() to handle opening the fire lock.
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:198]: Condensed, will be in the next commit.

File: code/game/machinery/doors/firedoor.dm [hunk 3]
```diff
@@ -21,14 +22,45 @@
...
+
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
+
+/obj/machinery/door/firedoor/Destroy()
+	remove_from_areas()
+	UnregisterSignal(src, COMSIG_MERGER_ADDING)
+	UnregisterSignal(src, COMSIG_MERGER_REMOVING)
+	return ..()
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:63]: It's good practice to do for components and elements, but it's not something you strictly have to do. any dangling signals will be cleared by datum/Destroy() (Unless it's a turf because turf refs are weird)

File: code/game/machinery/doors/firedoor.dm [hunk 4]
```diff
@@ -44,29 +76,142 @@
...
-	remove_from_areas()
-	affecting_areas.Cut()
-	return ..()
+	if(exposed_temperature > FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
+		start_activation_process(FIRELOCK_ALARM_TYPE_HOT)
+		return
+	if(exposed_temperature < BODYTEMP_COLD_DAMAGE_LIMIT)
+		start_activation_process(FIRELOCK_ALARM_TYPE_COLD)
+		return
+	activate()//Fallback. If firelocks regularly trigger themselves as generic, we need to do some bugfixing
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:146]: I'd prefer an error condition, though that may be mute because atmos sensitive isn't really a great solution for this

File: code/game/machinery/doors/firedoor.dm [hunk 5]
```diff
@@ -44,29 +76,142 @@
...
+ * This proc will call activate() on every fire lock (including us) listed
+ * in the merge group datum. Returns without doing anything if our alarm_type
+ * was already set, as that means that we're already active.
+ *
+ * Arguments:
+ * code should be one of three defined alarm types, or can be not supplied. Will dictate the color of the fire alarm lights, and defults to "firelock_alarm_type_generic"
+ */
+/obj/machinery/door/firedoor/proc/start_activation_process(code = FIRELOCK_ALARM_TYPE_GENERIC)
+	if(alarm_type)
+		return //We're already active
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:160]: start_activation_process is only called by atmos_expose right? and that already checks for alarm_type

File: code/game/machinery/doors/firedoor.dm [hunk 6]
```diff
@@ -44,29 +76,142 @@
...
+ * this fire lock.
+ */
+/obj/machinery/door/firedoor/proc/activate(code = FIRELOCK_ALARM_TYPE_GENERIC)
+	SIGNAL_HANDLER
+	if(alarm_type)
+		return //Already active
+	alarm_type = code
+	for(var/area/place in affecting_areas)
+		LAZYADD(place.active_firelocks, src)
+		if(LAZYLEN(place.active_firelocks) == 1) //if we're the first to activate in this particular area
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:183]: Why not handle fire alarm stuff by reacting to gas changes on the alarm itself? or is this to support manual use

File: code/game/machinery/doors/firedoor.dm [hunk 7]
```diff
@@ -44,29 +76,142 @@
...
+			for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+				fire_panel.set_status()
+			if(place == my_area)
+				place.alarm_manager.send_alarm(ALARM_FIRE, place) //We'll limit our reporting to just the area we're on. If the issue affects bordering areas, they can report it themselves
+	update_icon() //Sets the lights even if the door doesn't move.
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:188]: Oh I see what you mean

File: code/game/machinery/doors/firedoor.dm [hunk 8]
```diff
@@ -209,47 +362,62 @@
...
-	. += density ? "welded" : "welded_open"
+	if(welded)
+		. += density ? "welded" : "welded_open"
+	if(alarm_type && powered())
+		if(!warn_lights)
+			warn_lights = new()
+		warn_lights.icon = icon
+		warn_lights.icon_state = "[(obj_flags & EMAGGED) ? "firelock_alarm_type_emag" : alarm_type]"
+		warn_lights.plane = ABOVE_LIGHTING_PLANE
+		add_overlay(warn_lights)
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:375]: you should look at how fire alarms do glowing, it's cleaner then this and makes use of the emissive plane, which can be blocked

File: code/game/machinery/firealarm.dm [hunk 9]
```diff
@@ -115,34 +132,14 @@
...
-
-/obj/machinery/firealarm/atmos_end()
-	if(!detecting)
-		return
-	if(triggered)
-		triggered = FALSE
-		myarea.triggered_firealarms -= 1
-		update_appearance()
+	for(var/obj/machinery/door/firedoor/firelock in my_area.firedoors)
+		firelock.emag_act(user)
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/firealarm.dm:140]: What you've got here means you can break a whole area's worth of locks with one use on a fire alarm. That seems a bit much

File: code/game/machinery/doors/firedoor.dm [hunk 10]
```diff
@@ -44,29 +76,142 @@
...
+ *
+ * Arguments:
+ * code should be one of three defined alarm types, or can be not supplied. Will dictate the color of the fire alarm lights, and defults to "firelock_alarm_type_generic"
+ */
+/obj/machinery/door/firedoor/proc/start_activation_process(code = FIRELOCK_ALARM_TYPE_GENERIC)
+	if(alarm_type)
+		return //We're already active
+	var/datum/merger/merge_group = GetMergeGroup(merger_id, merger_typecache)
+	for(var/obj/machinery/door/firedoor/buddylock as anything in merge_group.members)
+		INVOKE_ASYNC(buddylock, .proc/activate, code)
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:163]: IIRC, firelocks were closing one at a time, though now that the correct_state() proc is called via async this one may not be needed.

File: code/game/machinery/doors/firedoor.dm (hunk 6 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:183]: It's a list of all active firelocks in an area. Fire alarms need to know if the number is greater than zero for their icon to be correct, and firelocks will avoid clearing the fire alert status of their home area if there's still an active firelock in the list after they remove themselves when resetting.

File: code/game/machinery/doors/firedoor.dm (hunk 7 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:188]: Next commit will specify "door lights" in the comment.

File: code/game/machinery/doors/firedoor.dm [hunk 11]
```diff
@@ -44,29 +76,142 @@
...
+/obj/machinery/door/firedoor/should_atmos_process(datum/gas_mixture/air, exposed_temperature)
+	return world.time > detect_cooldown && (exposed_temperature > FIRE_MINIMUM_TEMPERATURE_TO_EXIST || exposed_temperature < BODYTEMP_COLD_DAMAGE_LIMIT) && !(obj_flags & EMAGGED) && !machine_stat
+
+/obj/machinery/door/firedoor/atmos_expose(datum/gas_mixture/air, exposed_temperature)
+	if(obj_flags & EMAGGED)
+		return
+	for(var/area/place in affecting_areas)
+		if(!place.fire_detect) //if any area is set to disable detection
+			return
+	if(alarm_type)
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:137]: Hey, it's now tomorrow + 10 days.
atmos_sensitive was an element I created to make dealing with code that wants to process while some atmos related check is true as easy as possible.
I don't think it fits this usecase, since you don't actually need to process. The environment is either inhospitable or it's not, right?

File: code/game/machinery/firealarm.dm (hunk 9 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/firealarm.dm:140]: It's the only sane way I know to replicate the old emag function with what amounts to a many-to-many firelock/firealarm relationship we have going. The upside is that fire locks are fairly easy to find, and the best alternative I can think of is keeping track of emagged fire alarms (which would be harder to find and remove). Plus, due to how emagging works, you can't emag the same firelock twice, so if someone replaces the firelocks you gotta find another alarm if you want to break them all again.

File: code/game/machinery/doors/firedoor.dm [hunk 12]
```diff
@@ -44,29 +76,142 @@
...
+	UnregisterSignal(old_merger, COMSIG_MERGER_REFRESH_COMPLETE)
 
+/// Handles the firelocks to register only one signal per group
+/obj/machinery/door/firedoor/proc/MergerRefreshComplete(datum/merger/merger, list/leaving_members, list/joining_members)
+
+///////////////////////////////////////////////////////////////////
+// End of Merger
+
+/**
+ * Calculates what areas we should worry about.
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: This has been removed.

File: code/game/machinery/doors/firedoor.dm (hunk 3 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: These are actually just gone now

File: code/game/machinery/doors/firedoor.dm [hunk 13]
```diff
@@ -44,29 +76,142 @@
...
-			LAZYREMOVE(A.firedoors, src)
+		for(var/area/place in affecting_areas)
+			LAZYREMOVE(place.firedoors, src)
+			LAZYREMOVE(place.active_firelocks, src)
+			if(!LAZYLEN(place.active_firelocks)) //if we were the last firelock still active in this particular area
+				for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+					fire_panel.set_status()
+
+/obj/machinery/door/firedoor/should_atmos_process(datum/gas_mixture/air, exposed_temperature)
+	return world.time > detect_cooldown && (exposed_temperature > FIRE_MINIMUM_TEMPERATURE_TO_EXIST || exposed_temperature < BODYTEMP_COLD_DAMAGE_LIMIT) && !(obj_flags & EMAGGED) && !machine_stat
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: Done

File: code/game/machinery/doors/firedoor.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: This proc was rewritten and this isn't a thing anymore.

File: code/game/machinery/doors/firedoor.dm (hunk 11 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: Updated to not use atmos sensitive

File: code/game/machinery/doors/firedoor.dm (hunk 10 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: Nevermind, these are no longer async

File: code/game/machinery/doors/firedoor.dm [hunk 14]
```diff
@@ -44,29 +76,142 @@
...
+	SIGNAL_HANDLER
+	alarm_type = null
+	for(var/area/place in affecting_areas)
+		LAZYREMOVE(place.active_firelocks, src)
+		if(!LAZYLEN(place.active_firelocks)) //if we were the last firelock still active in this particular area
+			for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+				fire_panel.set_status()
+	detect_cooldown = world.time + DETECT_COOLDOWN_STEP_TIME
+	update_icon() //Sets the lights even if the door doesn't move.
+	INVOKE_ASYNC(src, .proc/correct_state)
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: Done, asyncs are only in correct_state() now.

File: code/game/machinery/doors/firedoor.dm [hunk 15]
```diff
@@ -209,47 +362,62 @@
 	icon_state = "[base_icon_state]_[density ? "closed" : "open"]"
 
 /obj/machinery/door/firedoor/update_overlays()
+	cut_overlays()
```
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:351]: Apparently not, when I tested it without this line, the overlay doesn't get removed.

File: code/game/machinery/doors/firedoor.dm [hunk 16]
```diff
@@ -272,6 +428,21 @@
 
 	AddElement(/datum/element/connect_loc, loc_connections)
 
+/obj/machinery/door/firedoor/update_overlays()
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/firedoor.dm:431]: why is this being overriden twice?

Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:431]: mistakes were made

File: code/game/machinery/doors/firedoor.dm (hunk 15 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: This was reworked anyway.

File: code/game/machinery/doors/firedoor.dm (hunk 8 above)
[the remark this answers is not in the document]
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: Done

File: code/modules/antagonists/traitor/equipment/Malf_Modules.dm [hunk 17]
```diff
@@ -701,6 +701,8 @@ GLOBAL_LIST_INIT(malf_modules, subtypesof(/datum/ai_module))
 		if(!is_station_level(F.z))
 			continue
 		F.obj_flags |= EMAGGED
+		for(var/obj/machinery/door/firedoor/firelock in F.my_area.firedoors)
+			firelock.obj_flags |= EMAGGED
```
Comment (LemonInTheDark) [inline code/modules/antagonists/traitor/equipment/Malf_Modules.dm:705]: Should this still be done for disabling thermal sensors?

File: code/game/machinery/doors/firedoor.dm [hunk 18]
```diff
@@ -333,14 +530,14 @@
 	icon_state = "[base_icon_state][constructionStep]"
 	return ..()
 
-/obj/structure/firelock_frame/attackby(obj/item/C, mob/user)
+/obj/structure/firelock_frame/attackby(obj/item/attacking_object, mob/user)
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/firedoor.dm:533]: Thank you

File: code/game/machinery/doors/firedoor.dm [hunk 19]
```diff
@@ -21,14 +22,62 @@
...
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
+
+/**
+ * Sets the offset for the warning lights.
+ *
+ * Used for special firelocks with light overlays that don't line up to their sprite.
+ */
+/obj/machinery/door/firedoor/proc/set_offset()
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/firedoor.dm:75]: update_offset maybe? set implies it accepts an argument to set the thing to

File: code/game/machinery/doors/firedoor.dm [hunk 20]
```diff
@@ -160,6 +313,8 @@
 
 	if(density)
 		open()
+		if(alarm_type)
+			addtimer(CALLBACK(src, .proc/correct_state), 2 SECONDS, TIMER_UNIQUE)
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/firedoor.dm:317]: Why this two second timer?

File: code/game/machinery/doors/firedoor.dm [hunk 21]
```diff
@@ -44,29 +76,142 @@
...
+	if(alarm_type)
+		return //Already active
+	alarm_type = code
+	for(var/area/place in affecting_areas)
+		LAZYADD(place.active_firelocks, src)
+		if(LAZYLEN(place.active_firelocks) == 1) //if we're the first to activate in this particular area
+			for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+				fire_panel.set_status()
+			if(place == my_area)
+				place.alarm_manager.send_alarm(ALARM_FIRE, place) //We'll limit our reporting to just the area we're on. If the issue affects bordering areas, they can report it themselves
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:192]: If/when we make locks self resolve, this should come from the lock itself rather then the area.

File: code/game/machinery/doors/firedoor.dm [hunk 22]
```diff
@@ -44,29 +93,131 @@
...
+			for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
+				fire_panel.set_status()
+	COOLDOWN_START(src, detect_cooldown, DETECT_COOLDOWN_STEP_TIME)
+	update_icon() //Sets the door lights even if the door doesn't move.
+	correct_state()
+
+/obj/machinery/door/firedoor/emag_act(mob/user, obj/item/card/emag/doorjack/digital_crowbar)
+	if(!istype(digital_crowbar))
+		return //emags get out
+	if(obj_flags &= EMAGGED || !user || digital_crowbar.charges < 1)
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/firedoor.dm:216]: why are you doing &=?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Looking a lot better

File: code/modules/antagonists/traitor/equipment/Malf_Modules.dm (hunk 17 above)
Comment (zxaber) [reply code/modules/antagonists/traitor/equipment/Malf_Modules.dm:705]: Yes. I've redone how emagging works; emagging a fire alarm now just disables the noise. A doorjack is needed to emag firelocks, and that will disable the firelock's detection and put it into a state where it will never close except when forced (via silicon command or crowbar).

That said, I probably need to make sure the malf module runs all the emag code now, yeah.

File: code/game/machinery/doors/firedoor.dm (hunk 20 above)
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm:317]: This is what makes a firelock close if it was crowbar'd open while active now. Rather than the fire alarm essentially pulsing all firelocks in the area to close every set number of seconds, a firelock will now close exactly two seconds after being opened.

The `correct_state()` proc will observe the `alarm_type` var, and will refuse to close if the alarm state has cleared within those two seconds.

(This is when the firelock is just opened quickly, not when it's being held open. There's a separate block of code to handle people walking away after holding a firelock open.)

File: code/modules/antagonists/traitor/equipment/Malf_Modules.dm (hunk 17 above)
Comment (zxaber) [reply code/modules/antagonists/traitor/equipment/Malf_Modules.dm]: It now calls the firelock's emag_act proc instead. Also, it finds firelocks by looking through the GLOB.machines list, rather than checking the list of firelocks in each fire alarm from the prior loop (which I suspect will be faster solely due to not dealing with duplicate entries).

File: code/game/machinery/doors/firedoor.dm (hunk 22 above)
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: Typo. Or maybe was dumb. I don't remember. Has been fixed.

File: code/game/machinery/doors/firedoor.dm (hunk 19 above)
Comment (zxaber) [reply code/game/machinery/doors/firedoor.dm]: Changed to `adjust_lights_starting_offset()`.

File: code/game/machinery/doors/firedoor.dm [hunk 23]
```diff
@@ -21,14 +22,45 @@
...
+	RegisterSignal(src, COMSIG_MERGER_ADDING, .proc/MergerAdding)
+	RegisterSignal(src, COMSIG_MERGER_REMOVING, .proc/MergerRemoving)
+	if(!merger_typecache)
+		merger_typecache = typecacheof(/obj/machinery/door/firedoor)
+
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/door/firedoor/LateInitialize()
+	. = ..()
+	GetMergeGroup(merger_id, allowed_types = typecacheof(/obj/machinery/door/firedoor))
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/doors/firedoor.dm:68]: I guess to be more clear.
We generate merger_typecache so it can be used here, because typecacheof is a reasonably expensive proc, and it's best to just use it once.

File: code/game/machinery/firealarm.dm [hunk 24]
```diff
@@ -49,17 +46,36 @@
...
+ * the alarm sound based on the my_area.fire var.
+ */
+/obj/machinery/firealarm/proc/set_status()
+	if(my_area.fire || LAZYLEN(my_area.active_firelocks))
+		soundloop.start()
+		set_light(l_power = 0.8)
+	else
+		soundloop.stop()
+		set_light(l_power = 0)
+	update_icon()
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/game/machinery/firealarm.dm]: https://github.com/tgstation/tgstation/pull/62055#discussion_r737158923
Would apprichaite a response here

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Fix the merger_typecache thing and we can jive.
I've still got some issues I think we need to handle, some firelocks are placed in windows for instance, so detecting based on their tile is a bit messy, but I am willing to handle that later

Sorry for the huge delay on this.

**Comment by LemonInTheDark** [comment]
Testmerged on sybil

**Comment by Mooshimi** [comment]
Um... why not add crowbar to survival boxes.?

**Comment by zxaber** [comment]
Because that's an AI nerf and so not my style

(I main a role that starts with a crowbar. Having a crowbar doesn't make firelock hell any more fun. The point here is that firelock hell should happen less often.)

File: code/game/machinery/firealarm.dm [hunk 25]
```diff
@@ -32,31 +30,49 @@
...
+ * the alarm sound based on the state of an area variable.
+ */
+/obj/machinery/firealarm/proc/set_status()
+	if( (my_area.fire || LAZYLEN(my_area.active_firelocks)) && !(obj_flags & EMAGGED) )
+		soundloop.start()
+		set_light(l_power = 0.8)
+	else
+		soundloop.stop()
+		set_light(l_power = 0)
+	update_icon()
```
Comment (LemonInTheDark) [inline code/game/machinery/firealarm.dm:74]: I wonder if it would be better to do the sound stuff based on the fire alarm's own state (triggered, in a breach/fire), rather then the area's, so it doesn't layer horribly

Comment (zxaber) [reply code/game/machinery/firealarm.dm:74]: For what it's worth, I cut the sound file down to two beeps, but left the loop time what it was before. It's now about 75% silence and is way better even if you have different fire alarms nearby on different loop timers.

**Comment by zxaber** [comment]
Update:
- Fire locks will check the `alarm_type` var before and after a close or open sleep, to fix an issue I saw on live where rapidly activating and reseting fire alarms would leave fire locks in the wrong state.

- Emagged firelocks will now still trigger an alarm and set their own (and neighbor's) alarm_type if they encounter fire or cold, but will fail to close while displaying the emag light pattern. This should help make it clear that the fire lock was tampered with (rather than not closing due to a bug).

**Comment by cacogen** [comment]
Just wanted to address concerns that optimising firelocks would stop atmospheric technicians from having anything to do

File: code/game/machinery/doors/firedoor.dm [hunk 26]
```diff
@@ -82,7 +255,13 @@
 
 /obj/machinery/door/firedoor/power_change()
 	. = ..()
-	INVOKE_ASYNC(src, .proc/latetoggle)
+	update_icon()
+	correct_state()
+	if(machine_stat & NOPOWER)
+		soundloop.volume = 0 //This is the easiest way to stop an alarm when we lose power without having to keep track of which firelock was playing the alarm sound.
```
Comment (LemonInTheDark) [inline code/game/machinery/doors/firedoor.dm:261]: Why is this needed? Couldn't you just .stop and .start them?
